schematic/store/synapse_tracker.py

```diff
@@ -0,0 +1,140 @@
+"""This script is responsible for creating a 'pull through cache' class that can be
+added through composition to any class where Synapse entities might be used. The idea
+behind this class is to provide a mechanism such that if a Synapse entity is requested
+multiple times, the entity is only downloaded once. This is useful for preventing
+multiple downloads of the same entity, which can be time consuming."""
+from dataclasses import dataclass, field
+from typing import Dict, List, Union
+
+import synapseclient
+from synapseclient import Entity, File, Folder, Project, Schema
+
+
+@dataclass
+class SynapseEntiyTracker:
+    """The SynapseEntiyTracker class handles tracking synapse entities throughout the
+    lifecycle of a request to schematic. It is used to prevent multiple downloads of
+    the same entity."""
+
+    synapse_entities: Dict[str, Union[Entity, Project, File, Folder, Schema]] = field(
+        default_factory=dict
+    )
+    project_headers: Dict[str, List[Dict[str, str]]] = field(default_factory=dict)
+    """A dictionary of project headers for each user requested."""
+
+    def get(
+        self,
+        synapse_id: str,
+        syn: synapseclient.Synapse,
+        download_file: bool = False,
+        retrieve_if_not_present: bool = True,
+        download_location: str = None,
+        if_collision: str = None,
+    ) -> Union[Entity, Project, File, Folder, Schema]:
+        """Retrieves a Synapse entity from the cache if it exists, otherwise downloads
+        the entity from Synapse and adds it to the cache.
+
+        Args:
+            synapse_id: The Synapse ID of the entity to retrieve.
+            syn: A Synapse object.
+            download_file: If True, download the file.
+            retrieve_if_not_present: If True, retrieve the entity if it is not present in the cache.
+            download_location: The location to download the file to.
+            if_collision: The action to take if there is a collision when downloading the file.
+
+        Returns:
+            The Synapse entity.
+        """
+        entity = self.synapse_entities.get(synapse_id, None)
+
+        if entity is None or (download_file and not entity.path):
```

Comment:
I found the logic here a bit difficult to follow without a comment. Could you add: "Returns None if `retrieve_if_not_present` is False and the entity is not found in the cache" as a comment?

Reply:
Updated docstring to read:
```
        """Retrieves a Synapse entity from the cache if it exists, otherwise downloads
        the entity from Synapse and adds it to the cache.

        Args:
            synapse_id: The Synapse ID of the entity to retrieve.
            syn: A Synapse object.
            download_file: If True, download the file.
            retrieve_if_not_present: If True, retrieve the entity if it is not present 
                in the cache. If not found in the cache, and this is False, return None.
            download_location: The location to download the file to.
            if_collision: The action to take if there is a collision when downloading 
                the file. May be "overwrite.local", "keep.local", or "keep.both". A 
                collision occurs when a file with the same name already exists at the 
                download location.

        Returns:
            The Synapse entity if found. When retrieve_if_not_present is False and the
            entity is not found in the local cache, returns None. If
            retrieve_if_not_present is True and the entity is not found in the local
            cache, retrieve the entity from Synapse and add it to the cache.
        """
```